File: backend/Services/GeminiCvService.cs
using System.Text;
using System.Text.Json;
using System.Text.RegularExpressions;

namespace CareerMail.Api.Services;

public class GeminiCvService : IGeminiCvService
{
    private readonly HttpClient _httpClient;
    private readonly ILogger<GeminiCvService> _logger;

    private static readonly string[] CandidateModels = new[]
    {
        "gemini-2.5-flash",
        "gemini-1.5-flash",
        "gemini-3.5-flash-lite",
        "gemini-3.6-flash",
        "gemini-3.5-flash",
        "gemini-flash-latest"
    };

    // Fallback verified corporate career portals
    private static readonly Dictionary<string, string> VerifiedCompanyPortals = new(StringComparer.OrdinalIgnoreCase)
    {
        { "DeepMind", "https://deepmind.google/careers/" },
        { "Google", "https://careers.google.com/jobs/results/" },
        { "Revolut", "https://www.revolut.com/careers/" },
        { "Deliveroo", "https://careers.deliveroo.co.uk/" },
        { "Amazon", "https://amazon.jobs/" },
        { "Microsoft", "https://careers.microsoft.com/" },
        { "Monzo", "https://job-boards.greenhouse.io/monzo" },
        { "Monzo Bank", "https://job-boards.greenhouse.io/monzo" },
        { "Canonical", "https://job-boards.greenhouse.io/canonical" },
        { "Cloudflare", "https://job-boards.greenhouse.io/cloudflare" },
        { "GitLab", "https://job-boards.greenhouse.io/gitlab" },
        { "Datadog", "https://job-boards.greenhouse.io/datadog" },
        { "Figma", "https://job-boards.greenhouse.io/figma" },
        { "Elastic", "https://job-boards.greenhouse.io/elastic" },
        { "Reddit", "https://job-boards.greenhouse.io/reddit" },
        { "Spotify", "https://jobs.lever.co/spotify" },
        { "Palantir", "https://jobs.lever.co/palantir" },
        { "Palantir Technologies", "https://jobs.lever.co/palantir" },
        { "Bloomberg", "https://www.bloomberg.com/company/careers/" },
        { "AstraZeneca", "https://careers.astrazeneca.com/search-jobs" },
        { "BBC", "https://careerssearch.bbc.co.uk/" },
        { "Snyk", "https://snyk.io/careers/" },
        { "Checkout.com", "https://www.checkout.com/careers" },
        { "ARM", "https://careers.arm.com/" },
        { "Skyscanner", "https://www.skyscanner.net/jobs/" },
        { "Wise", "https://wise.jobs/" },
        { "Starling Bank", "https://www.starlingbank.com/careers/" },
        { "Meta", "https://www.metacareers.com/" },
        { "Apple", "https://jobs.apple.com/" },
        { "Netflix", "https://jobs.netflix.com/" }
    };

    public GeminiCvService(HttpClient httpClient, ILogger<GeminiCvService> logger)
    {
        _httpClient = httpClient;
        _httpClient.Timeout = TimeSpan.FromSeconds(15);
        _logger = logger;
    }

    public async Task<(bool Success, string Message)> TestApiKeyAsync(string apiKey)
    {
        if (string.IsNullOrWhiteSpace(apiKey))
        {
            return (false, "API key cannot be empty.");
        }

        string cleanKey = apiKey.Trim();

        try
        {
            var payload = new
            {
                contents = new[]
                {
                    new
                    {
                        parts = new[]
                        {
                            new { text = "Respond with 'OK' only." }
                        }
                    }
                }
            };

            string jsonString = JsonSerializer.Serialize(payload);
            string lastError = string.Empty;

            foreach (var model in CandidateModels)
            {
                using var cts = new CancellationTokenSource(TimeSpan.FromSeconds(6));
                try
                {
                    var url = $"https://generativelanguage.googleapis.com/v1beta/models/{model}:generateContent?key={cleanKey}";
                    using var request = new HttpRequestMessage(HttpMethod.Post, url)
                    {
                        Content = new StringContent(jsonString, Encoding.UTF8, "application/json")
                    };

                    var response = await _httpClient.SendAsync(request, cts.Token);
                    if (response.IsSuccessStatusCode)
                    {
                        return (true, "✓ Gemini Connected: API key is valid and responsive.");
                    }

                    var errorBody = await response.Content.ReadAsStringAsync();
                    lastError = $"{response.StatusCode}: {errorBody}";
                }
                catch (Exception ex)
                {
                    lastError = ex.Message;
                }
            }

            return (false, $"Gemini validation error: {lastError}");
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error testing Gemini API key");
            return (false, $"Connection error: {ex.Message}");
        }
    }

    public async Task<GeminiStructuredCvProfileDto?> ExtractStructuredCvAsync(string rawCvText, string apiKey)
    {
        if (string.IsNullOrWhiteSpace(apiKey) || string.IsNullOrWhiteSpace(rawCvText))
        {
            return null;
        }

        string cleanKey = apiKey.Trim();

        try
        {
            var prompt = $@"You are an expert technical recruiter and resume parsing system.
Extract all skills, technologies, job titles, and experience levels from the candidate's CV.
Return ONLY valid JSON matching this schema:
{{
  ""technicalSkills"": [""Python"", ""PyTorch"", ""SQL""],
  ""programmingLanguages"": [""Python"", ""C#""],
  ""frameworks"": [""React"", ""FastAPI""],
  ""databases"": [""PostgreSQL"", ""Redis""],
  ""cloudPlatforms"": [""AWS"", ""GCP""],
  ""tools"": [""Docker"", ""Git""],
  ""dataSkills"": [""Data Science"", ""Statistical Modeling""],
  ""aiMlSkills"": [""Machine Learning"", ""Deep Learning"", ""LLMs""],
  ""softSkills"": [""Communication"", ""Problem Solving""],
  ""jobTitles"": [""Data Scientist"", ""ML Engineer""],
  ""experienceYears"": 0,
  ""education"": [""BSc Computer Science""],
  ""normalisedSkills"": [""python"", ""pytorch"", ""sql""]
}}

Candidate CV Text:
{rawCvText.Trim()}";

            var payload = new
            {
                contents = new[]
                {
                    new
                    {
                        parts = new[]
                        {
                            new { text = prompt }
                        }
                    }
                },
                generationConfig = new
                {
                    responseMimeType = "application/json",
                    temperature = 0.1
                }
            };

            string jsonPayload = JsonSerializer.Serialize(payload);
            string? rawResponse = null;

            foreach (var model in CandidateModels)
            {
                using var cts = new CancellationTokenSource(TimeSpan.FromSeconds(10));
                try
                {
                    var url = $"https://generativelanguage.googleapis.com/v1beta/models/{model}:generateContent?key={cleanKey}";
                    using var request = new HttpRequestMessage(HttpMethod.Post, url)
                    {
                        Content = new StringContent(jsonPayload, Encoding.UTF8, "application/json")
                    };

                    var response = await _httpClient.SendAsync(request, cts.Token);
                    if (response.IsSuccessStatusCode)
                    {
                        rawResponse = await response.Content.ReadAsStringAsync();
                        break;
                    }
                }
                catch
                {
                    // Fallback to next model
                }
            }

            if (string.IsNullOrWhiteSpace(rawResponse)) return null;

            using var doc = JsonDocument.Parse(rawResponse);
            var root = doc.RootElement;

            if (!root.TryGetProperty("candidates", out var candidates) || candidates.GetArrayLength() == 0)
            {
                return null;
            }

            var firstCandidate = candidates[0];
            if (!firstCandidate.TryGetProperty("content", out var content) ||
                !content.TryGetProperty("parts", out var parts) ||
                parts.GetArrayLength() == 0)
            {
                return null;
            }

            var rawOutput = parts[0].GetProperty("text").GetString() ?? string.Empty;
            rawOutput = Regex.Replace(rawOutput, @"^```(?:json)?\s*", "", RegexOptions.Multiline);
            rawOutput = Regex.Replace(rawOutput, @"\s*```$", "", RegexOptions.Multiline).Trim();

            var options = new JsonSerializerOptions { PropertyNameCaseInsensitive = true };
            return JsonSerializer.Deserialize<GeminiStructuredCvProfileDto>(rawOutput, options);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Failed to extract structured CV with Gemini AI");
            return null;
        }
    }

    public async Task<string?> ResolveRealJobUrlAsync(string jobTitle, string companyName, string location, string? currentUrl, string? apiKey)
    {
        // 1. If it's already a direct Greenhouse, Lever, Remotive, RemoteOK, Jobicy ATS link, use directly
        if (!string.IsNullOrWhiteSpace(currentUrl))
        {
            if (currentUrl.Contains("greenhouse.io") ||
                currentUrl.Contains("lever.co") ||
                currentUrl.Contains("remotive.com") ||
                currentUrl.Contains("remoteok.com") ||
                currentUrl.Contains("jobicy.com") ||
                currentUrl.Contains("ashbyhq.com") ||
                currentUrl.Contains("workable.com"))
            {
                return currentUrl;
            }
        }

        // 2. If Gemini API key is provided, ask Gemini for the genuine official career portal / application URL
        if (!string.IsNullOrWhiteSpace(apiKey))
        {
            string cleanKey = apiKey.Trim();
            try
            {
                var prompt = $@"You are an expert technical careers and corporate job recruiter.
Determine the official, authentic company career portal or direct application URL for this job:
- Job Title: {jobTitle}
- Company: {companyName}
- Location: {location}
- Candidate Current URL: {currentUrl}

Return ONLY a JSON object:
{{
  ""realJobUrl"": ""https://...""
}}
Provide the genuine official HTTPS career portal URL (e.g. greenhouse.io, lever.co, workday, or company's official career portal like careers.google.com, amazon.jobs, etc.).";

                var payload = new
                {
                    contents = new[]
                    {
                        new
                        {
                            parts = new[]
                            {
                                new { text = prompt }
                            }
                        }
                    },
                    generationConfig = new
                    {
                        responseMimeType = "application/json",
                        temperature = 0.1
                    }
                };

                string jsonPayload = JsonSerializer.Serialize(payload);

                foreach (var model in CandidateModels)
                {
                    using var cts = new CancellationTokenSource(TimeSpan.FromSeconds(6));
                    try
                    {
                        var url = $"https://generativelanguage.googleapis.com/v1beta/models/{model}:generateContent?key={cleanKey}";
                        using var request = new HttpRequestMessage(HttpMethod.Post, url)
                        {
                            Content = new StringContent(jsonPayload, Encoding.UTF8, "application/json")
                        };

                        var response = await _httpClient.SendAsync(request, cts.Token);
                        if (response.IsSuccessStatusCode)
                        {
                            var respString = await response.Content.ReadAsStringAsync();
                            using var doc = JsonDocument.Parse(respString);
                            if (doc.RootElement.TryGetProperty("candidates", out var candidates) && candidates.GetArrayLength() > 0)
                            {
                                var cand = candidates[0];
                                if (cand.TryGetProperty("content", out var c) && c.TryGetProperty("parts", out var p) && p.GetArrayLength() > 0)
                                {
                                    var text = p[0].GetProperty("text").GetString() ?? "";
                                    text = Regex.Replace(text, @"^```(?:json)?\s*", "", RegexOptions.Multiline);
                                    text = Regex.Replace(text, @"\s*```$", "", RegexOptions.Multiline).Trim();

                                    using var parsedDoc = JsonDocument.Parse(text);
                                    if (parsedDoc.RootElement.TryGetProperty("realJobUrl", out var urlProp))
                                    {
                                        var foundUrl = urlProp.GetString();
                                        if (!string.IsNullOrWhiteSpace(foundUrl) && foundUrl.StartsWith("http", StringComparison.OrdinalIgnoreCase))
                                        {
                                            return foundUrl;
                                        }
                                    }
                                }
                            }
                            break;
                        }
                    }
                    catch
                    {
                        // Try next model
                    }
                }
            }
            catch (Exception ex)
            {
                _logger.LogWarning(ex, "Gemini URL resolver encountered an issue, falling back to verified portals");
            }
        }

        // 3. Fallback to verified official corporate career portals
        foreach (var (comp, portalUrl) in VerifiedCompanyPortals)
        {
            if (companyName.Contains(comp, StringComparison.OrdinalIgnoreCase))
            {
                return portalUrl;
            }
        }

        // 4. Default clean fallback
        if (!string.IsNullOrWhiteSpace(currentUrl)) return currentUrl;
        return $"https://www.google.com/search?q={Uri.EscapeDataString($"{companyName} {jobTitle} careers apply")}";
    }
}

File: backend/app/services/gemini_cv_service.py
import json
import logging
import re
import urllib.parse
import httpx
from app.config import settings

logger = logging.getLogger(__name__)


class GeminiCvService:
    GEMINI_MODELS = [
        "gemini-2.5-flash",
        "gemini-2.0-flash",
        "gemini-1.5-flash",
        "gemini-1.5-pro"
    ]

    async def test_api_key(self, api_key: str) -> tuple[bool, str]:
        if not api_key or not api_key.strip():
            return False, "API key cannot be empty"

        clean_key = api_key.strip()
        for model in self.GEMINI_MODELS:
            url = f"https://generativelanguage.googleapis.com/v1beta/models/{model}:generateContent?key={clean_key}"
            payload = {
                "contents": [{
                    "parts": [{"text": "Hello, please reply with 'OK'."}]
                }]
            }
            try:
                async with httpx.AsyncClient(timeout=8.0) as client:
                    resp = await client.post(url, json=payload)
                    if resp.status_code == 200:
                        return True, f"Gemini API key is valid and connected ({model})."
                    elif resp.status_code in (400, 403, 401):
                        err_msg = resp.json().get("error", {}).get("message", "Authentication failed")
                        return False, f"Invalid Gemini API key: {err_msg}"
            except Exception as e:
                logger.warning(f"Failed testing model {model}: {e}")

        return False, "Could not verify Gemini API key. Please check network connection or key validity."

    async def extract_cv_profile(self, cv_text: str, api_key: str | None = None) -> dict | None:
        key = api_key or settings.gemini_api_key
        if not key:
            return None

        prompt = (
            "Analyze the following CV / Resume text and extract structured candidate data strictly as JSON.\n"
            "Required JSON fields:\n"
            "- extractedSkills: list of strings (technical skills, frameworks, databases, cloud, methodologies)\n"
            "- targetRoles: list of strings (e.g. ['Software Engineer', 'Backend Developer'])\n"
            "- experienceYears: integer (estimated total years of professional tech experience)\n"
            "- educationLevel: string (e.g. 'Bachelor of Science in Computer Science' or 'Master's Degree')\n"
            "- preferredLocation: string (e.g. 'London, United Kingdom' or 'Remote')\n"
            "- isRemotePreferred: boolean\n\n"
            f"CV TEXT:\n{cv_text[:4000]}"
        )

        for model in self.GEMINI_MODELS:
            url = f"https://generativelanguage.googleapis.com/v1beta/models/{model}:generateContent?key={key.strip()}"
            payload = {
                "contents": [{
                    "parts": [{"text": prompt}]
                }],
                "generationConfig": {
                    "responseMimeType": "application/json"
                }
            }
            try:
                async with httpx.AsyncClient(timeout=12.0) as client:
                    resp = await client.post(url, json=payload)
                    if resp.status_code == 200:
                        data = resp.json()
                        text_resp = data.get("candidates", [{}])[0].get("content", {}).get("parts", [{}])[0].get("text", "")
                        return json.loads(text_resp)
            except Exception as e:
                logger.warning(f"Gemini CV extraction model {model} failed: {e}")

        return None

    def resolve_real_job_url(
        self,
        job_id: str | None,
        title: str | None,
        company: str | None,
        location: str | None,
        current_url: str | None
    ) -> str:
        clean_company = (company or "").strip()
        clean_title = (title or "").strip()
        clean_loc = (location or "United Kingdom").strip()

        # Known direct career portals
        company_portals = {
            "monzo": "https://monzo.com/careers",
            "revolut": "https://revolut.com/careers",
            "deepmind": "https://deepmind.google/careers",
            "google": "https://careers.google.com",
            "bloomberg": "https://careers.bloomberg.com",
            "bbc": "https://careers.bbc.co.uk",
            "amazon": "https://amazon.jobs",
            "microsoft": "https://careers.microsoft.com",
            "stripe": "https://stripe.com/careers",
            "deliveroo": "https://deliveroo.co.uk/careers",
            "spotify": "https://spotify.com/careers",
            "palantir": "https://palantir.com/careers",
            "canonical": "https://canonical.com/careers",
            "cloudflare": "https://cloudflare.com/careers",
            "gitlab": "https://gitlab.com/careers",
            "datadog": "https://datadoghq.com/careers",
            "snyk": "https://snyk.io/careers",
            "elastic": "https://elastic.co/careers",
            "reddit": "https://reddit.com/careers",
            "figma": "https://figma.com/careers"
        }

        for comp_key, portal_url in company_portals.items():
            if comp_key in clean_company.lower():
                return portal_url

        if current_url and current_url.startswith("http") and not any(k in current_url for k in ["example.com", "placeholder"]):
            return current_url

        query = urllib.parse.quote_plus(f"{clean_company} {clean_title} careers {clean_loc}")
        return f"https://www.google.com/search?q={query}"
